src/net_processing.cpp

```diff
@@ -1267,52 +1285,92 @@ void static ProcessGetBlockData(CNode* pfrom, const CInv& inv, CConnman* connman
     }
 }
 
+//! Determine whether or not a peer can request a transaction, and return it (or nullptr if not found or not allowed).
+CTransactionRef static FindTxForGetData(const CNode* peer, const uint256& txid, const std::chrono::seconds mempool_req, const std::chrono::seconds now)
+{
+    auto txinfo = mempool.info(txid);
+    if (txinfo.tx) {
+        // If a TX could have been INVed in reply to a MEMPOOL request,
+        // or is older than UNCONDITIONAL_RELAY_DELAY, permit the request
+        // unconditionally.
+        if ((mempool_req.count() && txinfo.nTime <= mempool_req.count()) || txinfo.nTime <= (now - UNCONDITIONAL_RELAY_DELAY).count()) {
+            return std::move(txinfo.tx);
+        }
+    }
+
+    {
+        LOCK(cs_main);
+
+        // Otherwise, the transaction must have been announced recently.
+        if (State(peer->GetId())->vRecentlyAnnouncedInvs.contains(txid)) {
+            // If it was, it can be relayed from either the mempool...
+            if (txinfo.tx) return std::move(txinfo.tx);
+            // ... or the relay pool.
+            auto mi = mapRelay.find(txid);
+            if (mi != mapRelay.end()) return mi->second;
+        }
+    }
+
+    return {};
+}
+
 void static ProcessGetData(CNode* pfrom, CConnman* connman, const std::atomic<bool>& interruptMsgProc)
 {
     AssertLockNotHeld(cs_main);
 
     std::deque<CInv>::iterator it = pfrom->vRecvGetData.begin();
     std::vector<CInv> vNotFound;
     const CNetMsgMaker msgMaker(pfrom->GetSendVersion());
-    {
-        LOCK(cs_main);
 
-        while (it != pfrom->vRecvGetData.end() && (it->type == MSG_TX || it->type == MSG_WITNESS_TX)) {
-            if (interruptMsgProc)
-                return;
-            // Don't bother if send buffer is too full to respond anyway
-            if (pfrom->fPauseSend)
-                break;
+    const std::chrono::seconds now = std::chrono::duration_cast<std::chrono::seconds>(std::chrono::system_clock::now().time_since_epoch());
```

Comment:
the difference is because we do not have the `GetTime` template. In bitcoin, `GetTime` was added in 2019. Also in the bitcoin master branch, it is marked as deprecated. So I didn't port and used the equivalent demo std library